### PR TITLE
Correct Node location of `regExpTemplateLiteral.reDisjunction`

### LINKED ISSUE #39457: [Task]: Giving semantic api support for regexp
### Description

When we pass a Template Expression Node which is a regexp we should get the correct TypeSymbol and Symbol for that.

function symbol() {
    var reg = re `<cursor>`;
}

### Describe your task(s)

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by sanjana**
Fixes #39457

## Remarks
Related to: #39509.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by LakshanWeerasinghe — CHANGES_REQUESTED** [review]
This is not working when we add an empty literal and gives Index out of range.

```
var reg = re ``
```

**Comment by LakshanWeerasinghe** [comment]
> This is not working when we add an empty literal and gives Index out of range.
> 
> ```
> var reg = re ``
> ```

The tests are failing due to this reason. Need to handle this case.

**Comment by sanjana** [comment]
> > This is not working when we add an empty literal and gives Index out of range.
> > ```
> > var reg = re ``
> > ```
> 
> The tests are failing due to this reason. Need to handle this case.

Fixed the empty regexp template literal case in https://github.com/ballerina-platform/ballerina-lang/pull/39466/commits/e15b6697d378dc2c68791741db813cfbc4b87300.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java [hunk 1]
```diff
@@ -5478,10 +5479,16 @@ private BLangRawTemplateLiteral createRawTemplateLiteral(NodeList<Node> members,
...
 
-    private BLangNode createRegExpTemplateLiteral(NodeList<Node> reSequences, Location location) {
+    private BLangNode createRegExpTemplateLiteral(NodeList<Node> reSequences, Location location, Token startBackTick,
+                                                  Token endBackTick) {
         BLangRegExpTemplateLiteral regExpTemplateLiteral =
                 (BLangRegExpTemplateLiteral) TreeBuilder.createRegExpTemplateLiteralNode();
-        regExpTemplateLiteral.reDisjunction = (BLangReDisjunction) createReDisjunctionNode(reSequences, location);
+        
+        Location reDisjunctionPos = reSequences.size() > 0 ? getPosition(reSequences.get(0),
+                reSequences.get(reSequences.size() - 1)) : getPosition(startBackTick, endBackTick);
```
Comment (LakshanWeerasinghe) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java:5488]:
```suggestion
        Location reDisjunctionPos = getPosition(startBackTick, endBackTick);
```
What happen if we use this approach? We should get the same result right?

Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java]: Updated in https://github.com/ballerina-platform/ballerina-lang/pull/39466/commits/49b17111ddcdae21a63a3f284309f903bd11b8a8.

**Review by LakshanWeerasinghe — APPROVED** [review]
LGTM!

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java [hunk 2]
```diff
@@ -5478,10 +5479,14 @@ private BLangRawTemplateLiteral createRawTemplateLiteral(NodeList<Node> members,
         return literal;
     }
 
-    private BLangNode createRegExpTemplateLiteral(NodeList<Node> reSequences, Location location) {
+    private BLangNode createRegExpTemplateLiteral(NodeList<Node> reSequences, Location location, Token startBackTick,
+                                                  Token endBackTick) {
         BLangRegExpTemplateLiteral regExpTemplateLiteral =
                 (BLangRegExpTemplateLiteral) TreeBuilder.createRegExpTemplateLiteralNode();
-        regExpTemplateLiteral.reDisjunction = (BLangReDisjunction) createReDisjunctionNode(reSequences, location);
+
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java:5486]:
```suggestion
```

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expressions_test.bal [hunk 3]
```diff
@@ -358,3 +358,10 @@ function fooFn() returns future<int> {
 function barFn() returns int {
     return 10;
 }
+
+function testRegexp() {
+    var reA = re `[a-z0-9]+`;
+    var reB = re ``;
+    var reC = re `a`;
+    var reD = re `[a|b]+c`;
```
Comment (pcnfernando) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expressions_test.bal:366]:
```suggestion
    _ = re `[a-z0-9]+`;
    _ = re ``;
    _ = re `a`;
    _ = re `[a|b]+c`;
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java (hunk 2 above)
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java]: Fixed in https://github.com/ballerina-platform/ballerina-lang/pull/39466/commits/1d6a475506892f8791a542f5b35d14d606218d7f.

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expressions_test.bal (hunk 3 above)
Comment (sanjana) [reply tests/ballerina-compiler-api-test/src/test/resources/test-src/expressions_test.bal]: Fixed in https://github.com/ballerina-platform/ballerina-lang/pull/39466/commits/1d6a475506892f8791a542f5b35d14d606218d7f.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java [hunk 4]
```diff
@@ -2457,7 +2457,8 @@ public BLangNode transform(TemplateExpressionNode expressionNode) {
             case RAW_TEMPLATE_EXPRESSION:
                 return createRawTemplateLiteral(expressionNode.content(), getPosition(expressionNode));
             case REGEX_TEMPLATE_EXPRESSION:
-                return createRegExpTemplateLiteral(expressionNode.content(), getPosition(expressionNode));
+                return createRegExpTemplateLiteral(expressionNode.content(), getPosition(expressionNode),
+                        expressionNode.startBacktick(), expressionNode.endBacktick());
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java:2461]:
```suggestion
                return createRegExpTemplateLiteral(expressionNode.content(), getPosition(expressionNode),
                        getPosition(expressionNode.startBacktick(), expressionNode.endBacktick()));
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java [hunk 5]
```diff
@@ -5478,10 +5479,13 @@ private BLangRawTemplateLiteral createRawTemplateLiteral(NodeList<Node> members,
         return literal;
     }
 
-    private BLangNode createRegExpTemplateLiteral(NodeList<Node> reSequences, Location location) {
+    private BLangNode createRegExpTemplateLiteral(NodeList<Node> reSequences, Location location, Token startBackTick,
+                                                  Token endBackTick) {
         BLangRegExpTemplateLiteral regExpTemplateLiteral =
                 (BLangRegExpTemplateLiteral) TreeBuilder.createRegExpTemplateLiteralNode();
-        regExpTemplateLiteral.reDisjunction = (BLangReDisjunction) createReDisjunctionNode(reSequences, location);
+        Location reDisjunctionPos = getPosition(startBackTick, endBackTick);
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java:5486]:
```suggestion
```

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/ExpressionTypeTestNew.java [hunk 6]
```diff
@@ -135,6 +136,23 @@ public void testArrayLiteral() {
         assertEquals(memberType.typeKind(), STRING);
     }
 
+    @Test(dataProvider = "RegexpTemplateLiteralPosProvider")
+    public void testRegexpTemplateLiteralExpr(int sLine, int sCol, int eLine, int eCol) {
+        TypeSymbol type = getExprType(sLine, sCol, eLine, eCol);
+        assertEquals(type.typeKind(), TYPE_REFERENCE);
```
Comment (pcnfernando) [inline tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/ExpressionTypeTestNew.java:142]: Should we validate signature as well?

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expressions_test.bal [hunk 7]
```diff
@@ -358,3 +358,10 @@ function fooFn() returns future<int> {
 function barFn() returns int {
     return 10;
 }
+
+function testRegexp() {
+    _ = re `[a-z0-9]+`;
+    _ = re ``;
+    _ = re `a`;
+    _ = re `[a|b]+c`;
```
Comment (pcnfernando) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/expressions_test.bal:366]: Add two tests with expected type
regexp:RegExp and string:RegExp

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java (hunk 4 above)
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java]: Updated in https://github.com/ballerina-platform/ballerina-lang/pull/39466/commits/7ae121096651ef8776fb470d5f56096e846c08e6.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java [hunk 8]
```diff
@@ -5478,10 +5479,13 @@ private BLangRawTemplateLiteral createRawTemplateLiteral(NodeList<Node> members,
         return literal;
     }
 
-    private BLangNode createRegExpTemplateLiteral(NodeList<Node> reSequences, Location location) {
+    private BLangNode createRegExpTemplateLiteral(NodeList<Node> reSequences, Location location, Token startBackTick,
+                                                  Token endBackTick) {
```
[the remark this answers is not in the document]
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java]: Updated in https://github.com/ballerina-platform/ballerina-lang/pull/39466/commits/7ae121096651ef8776fb470d5f56096e846c08e6.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java (hunk 5 above)
Comment (sanjana) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeBuilder.java]: Updated in https://github.com/ballerina-platform/ballerina-lang/pull/39466/commits/7ae121096651ef8776fb470d5f56096e846c08e6.

**Comment by pcnfernando** [comment]
@sanjana Please check the build failrue

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/ExpressionTypeTestNew.java (hunk 6 above)
Comment (sanjana) [reply tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/ExpressionTypeTestNew.java:142]: Awaiting changes: https://github.com/ballerina-platform/ballerina-lang/pull/39509/commits/973652b34a52057d12e7ff7d57cf32da8b7fb56f from #39509.

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/expressions_test.bal (hunk 7 above)
Comment (sanjana) [reply tests/ballerina-compiler-api-test/src/test/resources/test-src/expressions_test.bal:366]: Added in https://github.com/ballerina-platform/ballerina-lang/pull/39466/commits/69edd7f4444b7efee9a2c0ea079479f2f0ca18b3.

**Comment by sanjana** [comment]
> @sanjana Please check the build failrue

Fixed in https://github.com/ballerina-platform/ballerina-lang/pull/39466/commits/b7b84bb88c9a9d6015fab8ad95a798aec787b9eb.

**Comment by pcnfernando** [comment]
> > @sanjana Please check the build failrue
> 
> Fixed in [b7b84bb](https://github.com/ballerina-platform/ballerina-lang/commit/b7b84bb88c9a9d6015fab8ad95a798aec787b9eb).

Build is still failing

**Review by pcnfernando — APPROVED** [review]

**Pull request merged**